common/fs/src/tail.rs

```diff
@@ -79,330 +83,396 @@ pub struct Tailer {
+        }
+    };
+    None
+}
 
-    /// Runs the main logic of the tailer, this can only be run once so Tailer is consumed
-    pub fn process<'a>(
-        &mut self,
-        buf: &'a mut [u8],
-    ) -> Result<impl Stream<Item = Result<LazyLineSerializer, CacheError>> + 'a, std::io::Error>
-    {
-        let events = {
-            match FileSystem::stream_events(self.fs_cache.clone(), buf) {
-                Ok(events) => events,
-                Err(e) => {
-                    warn!("tailer stream raised exception: {:?}", e);
-                    return Err(e);
-                }
+/// Runs the main logic of the tailer, this can only be run once so Tailer is consumed
+pub fn process(
+    state: Tailer,
+) -> Result<impl Stream<Item = Result<LazyLineSerializer, CacheError>>, std::io::Error> {
+    let events = {
+        match FileSystem::stream_events(state.fs_cache.clone()) {
+            Ok(events) => events,
+            Err(e) => {
+                warn!("tailer stream raised exception: {:?}", e);
+                return Err(e);
             }
-        };
-
-        debug!("Tailer starting with lookback: {:?}", self.lookback_config);
-
-        Ok(events
-            .enumerate()
-            .then({
-                let fs = self.fs_cache.clone();
-                let lookback_config = self.lookback_config.clone();
-                let initial_offsets = self.initial_offsets.clone();
-                let event_times = self.event_times.clone();
-
-                move |(event_idx, (event_result, event_time))| {
-                    let fs = fs.clone();
-                    let lookback_config = lookback_config.clone();
-                    let initial_offsets = initial_offsets.clone();
-                    let event_times = event_times.clone();
-
-                    async move {
-                        match event_result {
-                            Err(err) => Some(futures::stream::iter(vec![Err(err)]).left_stream()),
-                            Ok(event) => {
-                                // debounce events
-                                // check event_time, if it's before the previous one
-                                let key_and_previous_event_time = match event {
-                                    Event::Write(key) => {
-                                        let event_times = event_times.lock().await;
-                                        Some((key, event_times.get(&key).cloned()))
-                                    }
-                                    _ => None,
-                                };
-
-                                // Need to check if the event is within buffer_length
-                                if let Some((_, Some((prev_event_idx, previous_event_time)))) =
-                                    key_and_previous_event_time
+        }
+    };
+
+    debug!("Tailer starting with lookback: {:?}", state.lookback_config);
+
+    Ok(events
+        .enumerate()
+        .then({
```

Comment:
I'm pretty sure if this .then was a .unfold we wouldn't need to Arc<Mutex<T>> the fs and then clone/capture it anymore, we could just pass it along as state... We might even be able to get rid of the RefCells around the data...

It would make the whole thing a lot less painful. I'll put in a ticket to investigate it